openxr/examples/vulkan.rs

```diff
@@ -382,6 +438,151 @@ fn main() {
+                                        | vk::MemoryPropertyFlags::HOST_COHERENT,
+                                )
+                    })
+                    .unwrap();
+                let uniform_memory = vk_device
+                    .allocate_memory(
+                        &vk::MemoryAllocateInfo::builder()
+                            .allocation_size(uniform_buffer_mem_requirements.size)
+                            .memory_type_index(memory_type_index),
+                        None,
+                    )
+                    .unwrap();
+
+                vk_device
+                    .bind_buffer_memory(uniform_buffer, uniform_memory, 0)
+                    .unwrap();
+                (uniform_buffer, uniform_memory)
+            })
+            .collect();
+
+        let descriptor_pool = vk_device
+            .create_descriptor_pool(
+                &vk::DescriptorPoolCreateInfo::builder()
+                    .pool_sizes(&[vk::DescriptorPoolSize::builder()
+                        .ty(vk::DescriptorType::UNIFORM_BUFFER)
+                        .descriptor_count(PIPELINE_DEPTH as _)
+                        .build()])
+                    .max_sets(PIPELINE_DEPTH),
+                None,
+            )
+            .unwrap();
+
+        let descriptor_sets = vk_device
+            .allocate_descriptor_sets(
+                &vk::DescriptorSetAllocateInfo::builder()
+                    .descriptor_pool(descriptor_pool)
+                    .set_layouts(&[descriptor_layout; PIPELINE_DEPTH as _]),
+            )
+            .unwrap();
+
+        for (i, descriptor_set) in descriptor_sets.iter().enumerate() {
+            vk_device.update_descriptor_sets(
+                &[vk::WriteDescriptorSet::builder()
+                    .dst_set(*descriptor_set)
+                    .dst_binding(0)
+                    .dst_array_element(0)
+                    .descriptor_type(vk::DescriptorType::UNIFORM_BUFFER)
+                    .buffer_info(&[*vk::DescriptorBufferInfo::builder()
+                        .buffer(uniform_buffers[i].0)
+                        .offset(0)
+                        .range((mem::size_of::<CameraTransform<f32>>() * 2) as _)])
+                    .build()],
+                &[],
+            )
+        }
+
+        // Copy vertex data into vertex buffer
+        {
+            vk_device
+                .bind_buffer_memory(vertex_buffer, vertex_memory, 0)
+                .unwrap();
+            let data_ptr = vk_device
+                .map_memory(
+                    vertex_memory,
+                    0,
+                    vertex_buffer_size as u64,
+                    vk::MemoryMapFlags::empty(),
+                )
+                .unwrap();
+            let mut align = ash::util::Align::new(
+                data_ptr,
+                mem::align_of::<f32>() as u64,
+                vertex_buffer_mem_requirements.size,
+            );
```

Comment:
`Align` isn't doing anything useful here.

Reply:
What do you recommend instead? I did it this way because this is what I saw in the ash examples. (and I changed it again to use `vertex_buffer_mem_requirements.alignment` for the second argument in the force push I just did but I suspect that may have been incorrect after doing some more research)

Reply:
You shouldn't need to do anything special at all. Memory allocated according to the requirements of a vertex buffer will necessarily satisfy the buffer's GPU-side alignment requirements. I believe the CPU-side alignment should also be implicitly satisfied. Writing data starting at any location other than the start of the buffer, or with a non-dense stride, would constitute a mismatch with the vertex attribute layout declared in the pipeline.

Reply:
(unfortunately the ash examples, and indeed most introductory vulkan examples, aren't great; there's a PR in need of review to improve them)

Reply:
Would it be sound to construct an `&mut [T]` from gpu memory or should I always go through ptr:;write?

Reply:
I see `Align` does create a slice but my concern is that the data in the buffer may not be valid for reads. I think I'll use `std::ptr::copy` for now.

Reply:
Creating a slice to uninitialized memory would be unsound, yeah. [`copy_nonoverlapping`](https://doc.rust-lang.org/std/ptr/fn.copy_nonoverlapping.html) should be preferred.

Reply:
Sounds good